### ISSUE TITLE
Thoughts on auto-generating changelog?

### ISSUE THREAD

**Issue by cyrus01337** [melonDS-emu/melonDS#2205]
I noticed that the most recent release (1.0rc at the time of writing) had a lot of changes, but there was understandably a lot more that could be reasonably tracked. What if the changelog could be automatically generated based on PRs?

It could be setup via GitHub Actions in a new workflow file and I'd be happy to assist with setting this up myself to reduce any potential friction for other collaborators moving forward. I didn't see any prior discussions on this when sifting through issues so hoping this is a first.

Thoughts?

**Comment by RSDuck** [comment]
Wouldn't that be just be basically this: https://github.com/melonDS-emu/melonDS/compare/0.9.5...1.0rc ?

I think one problem would be that the individual commits are sometimes not really understandable to outside people, as often multiple commits form a more abstract feature which is later improved or fixed. Back and forth changes like fixing regressions from the last release are also part of the commit log but are completely irrelevant to the version changelog.

The best solution would probably to maintain the changelog as one goes along with development. But we lack the discipline for that.

**Comment by cyrus01337** [comment]
# TL;DR
Yes.

Since others can't necessarily understand the commits, can we find a way to squash commits with an explanation given by the contributor(s) (could be done through a [PR request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository))?

# Long; Read
That would essentially be it, yes, which would mean rules would need to be enforced so that this becomes more habitual.

The problem with this approach in retrospect is that the sudden introduction of rules can feel a bit iron fist-like. In an ideal world features should be in a branch where all commits occur, then they can be merged into main and the commits squashed with a single commit message describing the feature.

How does development currently work within the repository? I notice there's a lot of branches but don't know enough about emulation, so one way that commits could be organised so that the changelog **can** be automated would be to ensure features are setup as PRs to main (or their own branch for testing purposes as the main branch performs automated distribution via pipeline) and the desired message for the commit to be set in the PR description, which can be merged trivially. Contributors can be notified through a [PR request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) that I'm also happy to contribute to.
